### PR TITLE
#2108 decouple bash-completion from xonsh

### PR DESCRIPTION
**Pull request by imomaliev**
Initial version of #2108 for discussions. It was pretty easy to split bash-completion from xonsh, but I had to copy `_quote_paths` implementation as well, and some parts of `platform` module for future users of `py-bash-completion`(https://github.com/imomaliev/bash-completion-python/tree/decoupling-xonsh). I also noticed that there are many functions like `bash_command` and `windows_bash_command` which are only used in completion, I left them as they are for now. 

I think maybe it is worth creating whole python interface for bash which will include not only completions but also other things like pipes, running commands etc. There is interesting project which partially does this(https://github.com/alexcouper/bash)

### CODE REVIEW

**Comment by scopatz** [comment]
Hi @imomaliev, can we also vend the `bash_completions` file here please?  It is OK if it is just a copy.  This makes it easier for us to distribute xonsh.  The main key is that the new `bash_completions` module won't import anything from xonsh

**Comment by imomaliev** [comment]
Hello @scopatz 
> can we also vend the bash_completions file here please
it is not a file it is the package https://github.com/imomaliev/bash-completion-python/tree/decoupling-xonsh, should I combine it to one module and attach it here?

> The main key is that the new bash_completions module won't import anything from xonsh
Yes, of course, there are no dependencies from `xonsh`, as I mentioned before I already use this package in my other project.

**Comment by scopatz** [comment]
I personally think a single module makes it easier to use in other projects, and with xonsh we get a speed boost on load if we do it that way.  This repo shows how to package up a single file project: https://github.com/xonsh/lazyasd

**Comment by imomaliev** [comment]
@scopatz added `xonsh/bash_completion.py`.

**Comment by laerus** [comment]
python3.6 under AppVeyor

```
FAILURES ===================================
_____________________________ test_script[case5] ______________________________
case = ("\ndef dummy(args, inn, out, err):\n    out.write('hey!')\n    return 0\n\ndef dummy2(args, inn, out, err):\n    s = inn.read()\n    out.write(s.upper())\n    return 0\n\naliases['d'] = dummy\naliases['d2'] = dummy2\nd | d2\n", 'HEY!', 0)
    @pytest.mark.parametrize('case', ALL_PLATFORMS)
    def test_script(case):
        script, exp_out, exp_rtn = case
        out, err, rtn = run_xonsh(script)
>       assert exp_out == out
E       assert 'HEY!' == 'xonsh: To log full traceb... 9] Bad file descriptor\n'
E         - HEY!
E         + xonsh: To log full traceback to a file set: $XONSH_TRACEBACK_LOGFILE = <filename>
E         + Traceback (most recent call last):
E         +   File "C:\projects\xonsh\xonsh\proc.py", line 1304, in run
E         +     r = self.f(self.args, sp_stdin, sp_stdout, sp_stderr, spec)
E         +   File "C:\projects\xonsh\xonsh\proc.py", line 1157, in proxy_four
E         +     return f(args, stdin, stdout, stderr)
E         +   File "9b98ebbba2f3b87fc13df8676598047b", line 7, in dummy2
E         + OSError: [Errno 9] Bad file descriptor
tests\test_integrations.py:156: AssertionError
```

File: xonsh/bash_completion.py [hunk 1]
```diff
@@ -0,0 +1,276 @@
+import os
```
Comment (scopatz) [inline xonsh/bash_completion.py]: I think this should be in the `completers` directory.

File: xonsh/bash_completion.py [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+import re
+import sys
+import platform
+import shlex
+import pathlib
+import subprocess
+
+
```
Comment (scopatz) [inline xonsh/bash_completion.py]: Another thing I have to be nice is to have `__version__ = 'x.y.z'` in these files, so you know if you copy is out of date.

**Comment by imomaliev** [comment]
@scopatz it fails on python 3.6
```
WARNING: 'RE_BACKGROUND' defined in multiple locations:
  xonsh.ansi_colors:9 (assign)
  xonsh.style_tools:25 (assign)
WARNING: 'Token' defined in multiple locations:
  xonsh.style_tools:28 (assign)
  xonsh.tokenize:252 (assign)
  collapsed 38 modules
Amalgamating xonsh.completers
  excluding None
WARNING: 'PATTERN_NEED_QUOTES' defined in multiple locations:
  xonsh.completers.bash_completion:91 (functiondef)
  xonsh.completers.path:13 (functiondef)
WARNING: '_quote_paths' defined in multiple locations:
  xonsh.completers.bash_completion:123 (functiondef)
  xonsh.completers.path:128 (functiondef)
WARNING: '_quote_to_use' defined in multiple locations:
  xonsh.completers.bash_completion:114 (functiondef)
  xonsh.completers.path:119 (functiondef)
WARNING: 'complete_from_bash' defined in multiple locations:
  xonsh.completers.bash:6 (functiondef)
  xonsh.completers.bash_completion:210 (functiondef)
Traceback (most recent call last):
  File "setup.py", line 344, in <module>
    main()
  File "setup.py", line 340, in main
    setup(**skw)
  File "/opt/python/3.6.0/lib/python3.6/distutils/core.py", line 148, in setup
    dist.run_commands()
  File "/opt/python/3.6.0/lib/python3.6/distutils/dist.py", line 955, in run_commands
    self.run_command(cmd)
  File "/opt/python/3.6.0/lib/python3.6/distutils/dist.py", line 974, in run_command
    cmd_obj.run()
  File "setup.py", line 187, in run
    amalgamate_source()
  File "setup.py", line 66, in amalgamate_source
    'xonsh.completers', 'xonsh.history', 'xonsh.prompt'])
  File "/home/travis/build/xonsh/xonsh/amalgamate.py", line 547, in main
    src = amalgamate(order, graph, pkg)
  File "/home/travis/build/xonsh/xonsh/amalgamate.py", line 462, in amalgamate
    lines = rewrite_imports(name, pkg, order, imps)
  File "/home/travis/build/xonsh/xonsh/amalgamate.py", line 411, in rewrite_imports
    raise RuntimeError(msg)
RuntimeError: Cannot amalgamate import of amalgamated module:

  from xonsh.completers import bash_completion
```
What is correct way to fix this?

File: xonsh/completers/bash.py [hunk 3]
```diff
@@ -1,125 +1,10 @@
...
-COMP_WORDS=({line})
-COMP_LINE={comp_line}
-COMP_POINT=${{#COMP_LINE}}
-COMP_COUNT={end}
-COMP_CWORD={n}
-$_func {cmd} {prefix} {prev}
-
-for ((i=0;i<${{#COMPREPLY[*]}};i++)) do echo ${{COMPREPLY[i]}}; done
-"""
+from xonsh.completers import bash_completion
```
Comment (scopatz) [inline xonsh/completers/bash.py]: This import should be `from completers.bash_completions import complete_from_bash`

File: xonsh/completers/bash.py [hunk 4]
```diff
@@ -1,125 +1,10 @@
...
 
 
 def complete_from_bash(prefix, line, begidx, endidx, ctx):
```
Comment (scopatz) [inline xonsh/completers/bash.py]: This name can't be the same as the name in bash_completions.

**Comment by scopatz** [comment]
Hi @imomaliev, I have made a couple of inline comments that should help.  Also, you can read about what amalgamate is doing here: https://github.com/xonsh/amalgamate#amalgamate

**Comment by scopatz** [comment]
Basically, because amalgamate shares a global context, it is not safe to reuse names. That explains the warnings.  The error is because it looks like a self-import.  Let me know if that doesn't clear things up.

**Comment by imomaliev** [comment]
@scopatz Thanks, for your guidance, fixed amalgamate issues. Now it fails with 
```
Checking out gh-pages
git checkout -b gh-pages --track doctr_remote/gh-pages
error: Your local changes to the following files would be overwritten by checkout:
	xonsh/completers/__init__.py
Please, commit your changes or stash them before you can switch branches.
Aborting

M	xonsh/completers/__init__.py
HEAD is now at cf424fa... Merge d50fe8aec5dd29b5cda68d916f0907c1877d1ce9 into 84b62841c046e4931859422c99b22013406caf7d
```

**Comment by gforsyth** [comment]
Hi @imomaliev -- that's a docs building error. The changes you've made will make amalgamate alter the `__init__.py` mentioned at install time. You can do a `make amalgamate` to force amalgamation without installing, then commit the changes to `xonsh/completers/__init__.py`.

**Comment by imomaliev** [comment]
@gforsyth Thanks, `make amalgamate` failed
```
File "/path/to/xonsh/xonsh/codecache.py", line 65, in run_compiled_code
    func(code, glb, loc)
  File "eed7610e553738a45c5cbd04dd5c8a03", line 1, in <module>
NameError: name 'setup' is not defined
make: *** [amalgamate] Error 1
```
But I made it work with running this in ipython.
```
In [1]: import setup

In [2]: setup.amalgamate_source()
```

@scopatz All checks are passing. I don't really like the result of decoupling, but it is working, and we can start from here.

**Comment by scopatz** [comment]
Thanks @imomaliev! Given that this woks, I am going to go ahead and merge it and I'll provide another PR to help clean it up.

**Comment by imomaliev** [comment]
@scopatz Thanks.

**Comment by scopatz** [comment]
No problem, see #2161

**Comment by scopatz** [comment]
@imomaliev since #2161 has been merged in, I think this is ready to be broken out into a separate repo.

**Comment by gforsyth** [comment]
https://github.com/xonsh/py-bash-completion

unless folks have a different naming preference

**Comment by scopatz** [comment]
Right @gforsyth, sorry I meant "ready to move" this over there.  I like the name.

**Comment by gforsyth** [comment]
No offense taken, I couldn't remember if I'd pinged @imomaliev about it

**Comment by imomaliev** [comment]
Hello @scopatz @gforsyth,
>  I think this is ready to be broken out into a separate repo.
Yes, I am currently very busy, hopefully will have time handle this on weekends.

**Comment by scopatz** [comment]
OK sounds good!

**Pull request merged**